From 90d5a7a42dfd26d6f1b3576a1d91b016657b02b1 Mon Sep 17 00:00:00 2001
From: Harshaka Narayana <harshaka.narayana@broadcom.com>
Date: Tue, 7 Jul 2026 09:52:48 -0700
Subject: [PATCH] vmxnet3: fix BUG_ON in vmxnet3_get_hdr_len() for Geneve
 packets

When rcd->tcp is set based on the inner TCP header of a
Geneve-encapsulated packet that exceeds MTU and is non-LRO, the driver
hits BUG_ON(hdr.ipv4->protocol != IPPROTO_TCP) in vmxnet3_get_hdr_len()
because the outer IPv4 protocol is UDP (Geneve), not TCP.

Convert this BUG_ON to return 0, letting the caller's existing hlen == 0
check skip the LRO path gracefully. Also convert the preceding
BUG_ON(gdesc->rcd.tcp == 0) assertion to return 0 defensively.

Fixes: 45dac1d6ea04 ("vmxnet3: Changes for vmxnet3 adapter version 2 (fwd)")
Signed-off-by: Harshaka Narayana <harshaka.narayana@broadcom.com>
Signed-off-by: Ronak Doshi <ronak.doshi@broadcom.com>
Signed-off-by: Sankararaman Jayaraman <sankararaman.jayaraman@broadcom.com>
---
 drivers/net/vmxnet3/vmxnet3_drv.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/net/vmxnet3/vmxnet3_drv.c b/drivers/net/vmxnet3/vmxnet3_drv.c
index 40522afc05320..f706cacfa70fd 100644
--- a/drivers/net/vmxnet3/vmxnet3_drv.c
+++ b/drivers/net/vmxnet3/vmxnet3_drv.c
@@ -1530,7 +1530,8 @@ vmxnet3_get_hdr_len(struct vmxnet3_adapter *adapter, struct sk_buff *skb,
 		struct ipv6hdr *ipv6;
 		struct tcphdr *tcp;
 	} hdr;
-	BUG_ON(gdesc->rcd.tcp == 0);
+	if (gdesc->rcd.tcp == 0)
+		return 0;
 
 	maplen = skb_headlen(skb);
 	if (unlikely(sizeof(struct iphdr) + sizeof(struct tcphdr) > maplen))
@@ -1547,7 +1548,9 @@ vmxnet3_get_hdr_len(struct vmxnet3_adapter *adapter, struct sk_buff *skb,
 		BUG_ON(hdr.eth->h_proto != htons(ETH_P_IP) &&
 		       hdr.veth->h_vlan_encapsulated_proto != htons(ETH_P_IP));
 		hdr.ptr += hlen;
-		BUG_ON(hdr.ipv4->protocol != IPPROTO_TCP);
+		if (hdr.ipv4->protocol != IPPROTO_TCP)
+			return 0;
+
 		hlen = hdr.ipv4->ihl << 2;
 		hdr.ptr += hdr.ipv4->ihl << 2;
 	} else if (gdesc->rcd.v6) {
-- 
2.34.1

